File: go/ie/pi.go
package ie

import (
	"regexp"

	"github.com/matheuslm7/br-validator/go/internal/shared"
)

// Verified against the official SEFAZ-PI "Roteiro de Crítica da Inscrição
// Estadual" (sintegra.gov.br/Cad_Estados/cad_PI.html), including its worked
// example (012345679). Format: 8 digits + 1 check digit.
var (
	piAllowedChars = regexp.MustCompile(`^[\d\s]+$`)
	piWeights      = []int{9, 8, 7, 6, 5, 4, 3, 2}
)

func piNormalize(value string) string {
	return shared.RemoveNonDigits(value)
}

func piIsValid(value string) bool {
	if !piAllowedChars.MatchString(value) {
		return false
	}

	v := piNormalize(value)

	if len(v) != 9 {
		return false
	}

	return shared.Mod11CheckDigit(shared.WeightedSum(v, piWeights)) == int(v[8]-'0')
}

func piFormat(value string) string {
	return piNormalize(value)
}

File: csharp/src/BrValidator/Ie/Pi.cs
using System.Text.RegularExpressions;
using BrValidator.Internal;

namespace BrValidator.Ie
{
    /// <summary>
    /// PI: verified against the official SEFAZ-PI "Roteiro de Crítica da
    /// Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_PI.html),
    /// including its worked example (012345679). Format: 8 digits + 1
    /// check digit.
    /// </summary>
    internal static class Pi
    {
        private static readonly Regex AllowedChars = new Regex(@"^[\d\s]+$", RegexOptions.Compiled);
        private static readonly int[] Weights = { 9, 8, 7, 6, 5, 4, 3, 2 };

        public static string Normalize(string value) => Shared.RemoveNonDigits(value);

        public static bool IsValid(string value)
        {
            if (!AllowedChars.IsMatch(value))
            {
                return false;
            }

            string v = Normalize(value);

            if (v.Length != 9)
            {
                return false;
            }

            return Mod11.Mod11CheckDigit(Mod11.WeightedSum(v, Weights)) == (v[8] - '0');
        }

        public static string Format(string value) => Normalize(value);
    }
}
